File: OCP/java/OCPBreak.java
import java.util.ArrayList;

// Product class
class Product {
    String name;
    int price;

    public Product(String name, int price) {
        this.name = name;
        this.price = price;
    }
}

// ShoppingCart handles products and total price
class ShoppingCart {
    private ArrayList<Product> products = new ArrayList<>();

    public void addProduct(Product product) {
        products.add(product);
    }

    public ArrayList<Product> getProducts() {
        return products;
    }

    public int calculateTotalPrice() {
        int total = 0;
        for (Product product : products) {
            total += product.price;
        }
        return total;
    }
}

// Printer class also prints invoice
class ShoppingCartPrinter {
    private ShoppingCart cart;

    public ShoppingCartPrinter(ShoppingCart cart) {
        this.cart = cart;
    }

    public void printInvoice() {
        System.out.println("\nInvoice:");
        for (Product product : cart.getProducts()) {
            System.out.println("Product Name: " + product.name);
            System.out.println("Product Price: " + product.price);
        }
        System.out.println("Total Price: " + cart.calculateTotalPrice());
    }
}

// Storage class has multiple responsibilities for different storage types
class ShoppingCartStorage {
    private ShoppingCart cart;

    public ShoppingCartStorage(ShoppingCart cart) {
        this.cart = cart;
    }

    public void saveToSQL() {
        System.out.println("\nSaving shopping cart to SQL DB");
    }

    public void saveToMongo() {
        System.out.println("\nSaving shopping cart to MongoDB");
    }

    public void saveToFile() {
        System.out.println("\nSaving shopping cart to file");
    }
}

// Main
public class OCPBreak {
    public static void main(String[] args) {
        ShoppingCart cart = new ShoppingCart();

        Product product1 = new Product("Chips", 10);
        Product product2 = new Product("Milk", 30);

        cart.addProduct(product1);
        cart.addProduct(product2);

        ShoppingCartPrinter printer = new ShoppingCartPrinter(cart);
        printer.printInvoice();

        ShoppingCartStorage storage = new ShoppingCartStorage(cart);
        storage.saveToSQL();
        storage.saveToMongo();
        storage.saveToFile();
    }
}

File: OCP/typescript/ocp_break.ts
class Product {
    public name: string;
    public price: number;

    constructor(name: string, price: number) {
        this.name = name;
        this.price = price;
    }
}

class ShoppingCart {
    private products: Product[];

    constructor() {
        this.products = [];
    }

    addProduct(product: Product): void {
        this.products.push(product);
    }

    getProducts(): Product[] {
        return this.products;
    }

    calculateTotalPrice(): number {
        let totalPrice = 0;

        for (const product of this.products) {
            totalPrice = totalPrice + product.price;
        }

        return totalPrice;
    }
}

class ShoppingCartStorage {
    private cart: ShoppingCart;

    constructor(cart: ShoppingCart) {
        this.cart = cart;
    }

    saveToSqlDb(): void {
        console.log("Saving shopping cart to SQL DB");
    }

    saveToMongoDb(): void {
        console.log("Saving shopping cart to Mongo DB");
    }

    saveToFile(): void {
        console.log("Saving shopping cart to file");
    }
}

const cart = new ShoppingCart();

cart.addProduct(new Product("Chips", 10));
cart.addProduct(new Product("Milk", 30));

const storage = new ShoppingCartStorage(cart);

storage.saveToSqlDb();
storage.saveToMongoDb();
